confidant/services/jwkmanager.py

```diff
@@ -22,15 +25,102 @@
 CA_SCHEMA = {
     'crt': {'type': 'string', 'required': True},
     'key': {'type': 'string', 'required': True},
-    'passphrase': {'type': 'string', 'required': True},
+    'passphrase': {'type': 'string', 'required': True, 'nullable': True},
     'kid': {'type': 'string', 'required': True},
 }
 
 
+class JwtCache(ABC):
+    @abstractmethod
+    def get_jwt(self, kid: str, requester: str, user: str) -> str:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def set_jwt(self, kid: str, requester: str, user: str, expiry: str,
+                jwt: str) -> None:
+        raise NotImplementedError()
+
+
+class LocalJwtCache(JwtCache):
+    def __init__(self) -> None:
+        # format of self._token_cache:
+        # {
+        #     "<kid>": {
+        #         "<downstream_requester>": {
+        #            "<requested_resource_id>": {
+        #                "expiry": 1234567890,
+        #                "jwt": "eyJpc19zZ...",
+        #         },
+        #         ...
+        #     },
+        #     ...
+        # }
+        # Example:
+        # {
+        #     "0h7R8..": {
+        #         "serviceA-staging-iad": {
+        #             "ServiceAA-staging-iad": {
+        #                 "expiry": 1234567890,
+        #                 "jwt": "eyJpc19zZX...",
+        #         },
+        #         ...
+        #     },
+        #     ...
+        # }
+        self._token_cache = {}
+
+    def get_jwt(self, kid: str, requester: str, user: str) -> str:
+        jwt_str = None
+
+        token_kv = self._token_cache.get(kid, {}).get(requester, {}).get(user)
+        if token_kv:
+            now = datetime.now(tz=timezone.utc)
+            when_created = token_kv['expiry'] - timedelta(
+                seconds=JWT_DEFAULT_JWT_EXPIRATION_SECONDS
+                )
+            elapsed_time = (now - when_created).total_seconds()
+            if elapsed_time < JWT_CACHING_TTL_SECONDS:
+                stats.incr('jwt.get_jwt.cache.hit')
+                jwt_str = token_kv['jwt']
+            else:
+                stats.incr('jwt.get_jwt.cache.expired')
+                del self._token_cache[kid][requester][user]
+        else:
+            stats.incr('jwt.get_jwt.cache.miss')
+        return jwt_str
+
+    def set_jwt(self, kid: str, requester: str, user: str, expiry: str,
+                jwt: str) -> None:
+        if kid not in self._token_cache:
+            self._token_cache[kid] = {}
+
+        if requester not in self._token_cache[kid]:
+            self._token_cache[kid][requester] = {}
```

Comment:
We can shove the entire thing as the dictionary key. it will make it the code simpler
`key = f'jwt:{kid}:{requester}:{user}'`

Reply:
good idea, i'll update